Fix issue where resizing an image doesn't always give an exact width/height

shssoichiro opened the pull request:
Also refactors out common resizing logic from `resize` and `resize_to_fill`

It seems that this was due to insufficient precision in the `f32` type which was causing rounding errors in some cases. `f64` seems to have enough precision to handle this accurately, at least in the cases I tested.

Closes #667

calum commented on src/dynimage.rs:
```diff
@@ -324,6 +302,25 @@ impl DynamicImage {
         }
     }
 
+    fn resize_dimensions(&self, nwidth: u32, nheight: u32, fill: bool) -> (u32, u32) {
```
I think it would be beneficial to have a unit test for this function. However, I'm not sure where it should go. There aren't many tests written at the bottom of this file.

bvssvni replied:
The bottom at the file is fine, I think. Could put it here: https://github.com/PistonDevelopers/image/blob/master/src/dynimage.rs#L764

shssoichiro commented:
Any update on this? I've added unit tests around the resizing logic.

bvssvni commented:
Looks good to me.

HeroicKatora commented:
Wouldn't it be better to use integer logic with `u64` instead? As in 
`iwidth / iheight < nwidht / nheight` <=> `iwhdth * nheight < nwidth*iheight`

`f64` with 53 bits of precision could be enough though. Compared to the old, the input arguments are surely represented without error and the division result is only be off by at most one `ulp`. Maybe this should be checked more thoroughly regardless.

shssoichiro commented:
I do like that better, if nothing else because integer math is generally faster. We still have to use float division to calculate the ratios, but can safely use u64's for calculating the new dimensions.

shssoichiro commented:
I also realized I could simplify two of the expressions e.g. `(width as u64 * nwidth as u64 / width as u64) as u32` to just `nwidth`. So I pushed another new commit to include that.

HeroicKatora commented:
Should we handle the case that calling with `fill = True` fails the cast to `u32`, i.e. the size got too big?

shssoichiro commented:
I agree. I've added tests around fill with and without overflow, and updated the code to handle the overflow case by bounding it to `u32::MAX`. I'm not sure that this code is the cleanest, but it does handle the edge cases.

shssoichiro commented:
I had to push a commit to fix a compile error if pnm feature is disabled; I was trying to pull this branch into a project where only png and jpeg are enabled, and encountered that. I realize it's not related to this issue, but I needed this branch to work.

bvssvni commented on src/dynimage.rs:
```diff
@@ -747,6 +725,31 @@ pub fn guess_format(buffer: &[u8]) -> ImageResult<ImageFormat> {
     )
 }
 
+fn resize_dimensions(width: u32, height: u32, nwidth: u32, nheight: u32, fill: bool) -> (u32, u32) {
```
Can you add some comments? Explain e.g. `fill`.

bvssvni commented:
Merging.
